### PR TITLE
HPCC-18435 Implement new versions of segment monitors

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

File: rtl/eclrtl/rtlnewkey.hpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    CMPeq = 0x02,
+    CMPgt = 0x04,
+    CMPmin = 0x08,      //minimum possible value
+    CMPmax = 0x10,      //maximum possible value
+    CMPle = CMPlt | CMPeq,
+    CMPge = CMPgt | CMPeq,
+    CMPminmask = CMPgt|CMPmin,
+    CMPmaxmask = CMPlt|CMPmax,
+};
+inline constexpr TransitionMask operator | (TransitionMask l, TransitionMask r) { return (TransitionMask)((unsigned)l | (unsigned)r); } \
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlnewkey.hpp]: Spurious trailing \

File: rtl/eclrtl/rtlnewkey.hpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    //starts searching from curRange (which is likely to match).  Uses a sequential search.
+    virtual int checkCandidateRange(const byte * field, unsigned & curRange) const = 0;
+
+    // Does this field match any range?
+    virtual bool matches(const byte * field) const = 0;
+
+    // Does this field match this particular range?
+    virtual bool matches(const byte * field, unsigned range) const = 0;
+
+    //Should this belong elsewhere?
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlnewkey.hpp]: This comment seems to be targeting a different audience to the others...

Comment (richardkchapman) [reply rtl/eclrtl/rtlnewkey.hpp]: The issue being I think that this method is just a passthrough to the associated type compare function, so why not provide a way to retrieve that typeInfo structure instead?

File: rtl/eclrtl/rtlfield.cpp [hunk 3]
```diff
@@ -1541,14 +1708,37 @@ size32_t RtlUnicodeTypeInfo::buildUtf8(ARowBuilder &builder, size32_t offset, co
...
 }
 
+size32_t RtlUnicodeTypeInfo::buildString(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, size32_t sizeInChars, const char *value) const
+{
+    if (!isFixedSize())
+    {
+        size32_t sizeInBytes = sizeInChars * sizeof(UChar);
+        builder.ensureCapacity(offset+sizeInBytes+sizeof(size32_t), queryName(field));
+        byte *dest = builder.getSelf()+offset;
+        rtlWriteInt4(dest, sizeInChars);  // NOTE - in chars!
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Should this be rtlWriteIntSize32t ?

File: rtl/eclrtl/rtlfield.cpp [hunk 4]
```diff
@@ -2240,6 +2526,48 @@ void RtlSetTypeInfo::readAhead(IRowDeserializerSource & in) const
     in.skip(thisLength);
 }
 
+//Should the parameters be const byte * - probably.
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Should this comment be removed?

File: rtl/eclrtl/rtlnewkey.hpp (hunk 1 above)
Comment (ghalliday) [reply rtl/eclrtl/rtlnewkey.hpp]: I'll move something from jstats and reuse that

**Comment by ghalliday** [comment]
@richardkchapman ready for first review

File: rtl/eclrtl/rtlfield.hpp [hunk 5]
```diff
@@ -267,6 +294,7 @@ struct ECLRTL_API RtlCharTypeInfo : public RtlTypeInfoBase
     virtual void getString(size32_t & resultLen, char * & result, const void * ptr) const override;
     virtual void getUtf8(size32_t & resultLen, char * & result, const void * ptr) const override;
     virtual __int64 getInt(const void * ptr) const override;
+    virtual int compare(const byte * left, const byte * right) const override;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.hpp]: Needs buildString/buildUtf8 too (so does RtlUnimplementedTypeInfo)

File: rtl/eclrtl/rtlfield.hpp [hunk 6]
```diff
@@ -493,6 +542,7 @@ struct ECLRTL_API RtlUnimplementedTypeInfo : public RtlTypeInfoBase
     virtual void getUtf8(size32_t & resultLen, char * & result, const void * ptr) const override;
     virtual __int64 getInt(const void * ptr) const override;
     virtual bool isScalar() const override { return false; }
+    virtual int compare(const byte * left, const byte * right) const override;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.hpp]: Needs buildString/buildUtf8 too

File: rtl/eclrtl/rtlnewkey.hpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef RTLNEWKEY_INCL
+#define RTLNEwKEY_INCL
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlnewkey.hpp]: Typo (lower case w)

File: rtl/eclrtl/rtlnewkey.cpp [hunk 8]
```diff
@@ -0,0 +1,2133 @@
...
+        }
+    }
+    ValueTransition(TransitionMask _mask) : mask(_mask)
+    {
+        dbgassertex(isMaximum() || isMinimum());
+    }
+
+    MemoryBuffer &serialize(const RtlTypeInfo & type, MemoryBuffer &mb) const
+    {
+        mb.append(mask);
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlnewkey.cpp]: This gives me a compile error:

/Users/rchapman/HPCC-Platform/rtl/eclrtl/rtlnewkey.cpp:156:12: error: call to member function 'append' is ambiguous

**Comment by ghalliday** [comment]
Fixed missing virtuals and append ambiguity.

File: rtl/eclrtl/eclhelper_dyn.cpp [hunk 9]
```diff
@@ -136,52 +136,7 @@ class ECLRTL_API CDynamicDiskReadArg : public CThorDiskReadArg
...
-            {
-                char startRange = *filter++;
-                if (startRange != '(' && startRange != '[')
-                    throw MakeStringException(0, "Invalid filter string: expected [ or ( at start of range");
-                // Now we expect a constant - type depends on type of field. Assume string or int for now
-                StringBuffer upperString, lowerString;
-                if (*filter=='\'')
-                {
-                    filter++;
-                    readString(lowerString, filter);
```
Comment (richardkchapman) [inline rtl/eclrtl/eclhelper_dyn.cpp]: The static readString function is now unused, I think, and should be deleted

File: rtl/eclrtl/rtlkey.cpp [hunk 10]
```diff
@@ -2317,6 +2318,42 @@ ECLRTL_API IStringSet *deserializeStringSet(MemoryBuffer &mb)
...
+
+protected:
+    IStringSet * set;
+    const RtlTypeInfo *fieldType;
+    size32_t minRecordSize;
+    MemoryBuffer lobuffer;
+    MemoryBuffer hibuffer;
+};
+
+void deserializeSet(IStringSet * set, size32_t minRecordSize, const RtlTypeInfo * fieldType, const char * filter)
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlkey.cpp]: Would IStringSet &set be better?

File: system/jlib/jstats.h [hunk 11]
```diff
@@ -252,12 +252,6 @@ class jlib_decl StatisticValueFilter
 
 //These could be template definitions, but that would potentially affect too many classes.
```
Comment (richardkchapman) [inline system/jlib/jstats.h]: This comment should be moved along with the code it referred to (or deleted)

File: rtl/eclrtl/eclhelper_dyn.cpp [hunk 12]
```diff
@@ -136,52 +136,7 @@ class ECLRTL_API CDynamicDiskReadArg : public CThorDiskReadArg
             MemoryBuffer lobuffer;
             MemoryBuffer hibuffer;
             Owned<IStringSet> filterSet = createStringSet(fieldSize);
```
Comment (richardkchapman) [inline rtl/eclrtl/eclhelper_dyn.cpp]: Should switch this to use the new ValueSet I would think (maybe can't for IndexRead case, but no reason not to for diskread is there?)

File: rtl/eclrtl/rtlfield.cpp [hunk 13]
```diff
@@ -2073,6 +2416,18 @@ size32_t RtlRecordTypeInfo::buildNull(ARowBuilder &builder, size32_t offset, con
     return offset;
 }
 
+size32_t RtlRecordTypeInfo::buildString(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, size32_t size, const char *value) const
+{
+    UNIMPLEMENTED;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Do you ever intend to implement/need? If not should probably be throwUnexpected()

File: rtl/eclrtl/rtlfield.cpp [hunk 14]
```diff
@@ -2073,6 +2416,18 @@ size32_t RtlRecordTypeInfo::buildNull(ARowBuilder &builder, size32_t offset, con
...
+    return 0;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Unreachable code

File: rtl/eclrtl/rtlfield.cpp [hunk 15]
```diff
@@ -2090,8 +2445,27 @@ __int64 RtlRecordTypeInfo::getInt(const void * ptr) const
     return 0;
 }
 
+int RtlRecordTypeInfo::compare(const byte * left, const byte * right) const
+{
+    return compareFields(fields, left, right);
+}
+
+
+
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Excess of blank lines?

File: rtl/eclrtl/rtlfield.cpp (hunk 13 above)
Comment (richardkchapman) [reply rtl/eclrtl/rtlfield.cpp]: (Could actually be useful for my dumpkey code...)

File: rtl/eclrtl/rtlfield.cpp [hunk 16]
```diff
@@ -2617,14 +3104,60 @@ size32_t RtlDictionaryTypeInfo::deserialize(ARowBuilder & builder, IRowDeseriali
...
+            {
+                if (countLeft == countRight)
+                    return 0;
+                if (row == countLeft)
+                    return -1;
+                else
+                    return +1;
+            }
+
+            //MORE: Is this correct
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: I don't know but should probably decide before merging the code!

File: rtl/eclrtl/rtlfield.cpp [hunk 17]
```diff
@@ -2617,14 +3104,60 @@ size32_t RtlDictionaryTypeInfo::deserialize(ARowBuilder & builder, IRowDeseriali
...
+            }
+            int rc = child->compare(curLeft, curRight);
+            if (rc != 0)
+                return rc;
+            row++;
+        }
+    }
+    else
+    {
+        UNIMPLEMENTED;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Will it be? When is it going to be needed? Is there a Jira to cover it?

File: rtl/eclrtl/rtlfield.cpp [hunk 18]
```diff
@@ -2639,6 +3172,16 @@ size32_t RtlIfBlockTypeInfo::size(const byte * self, const byte * selfrow) const
     return 0;
 }
 
+size32_t RtlIfBlockTypeInfo::buildString(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, size32_t size, const char *value) const
+{
+    throwUnexpected();
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Some inconsistency about whether we are using rtlFailUnexpected() or throwUnexpected()

File: rtl/eclrtl/rtlfield.cpp [hunk 19]
```diff
@@ -2782,6 +3394,18 @@ size32_t RtlUnimplementedTypeInfo::size(const byte * self, const byte * selfrow)
     return 0;
 }
 
+size32_t RtlUnimplementedTypeInfo::buildString(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, size32_t size, const char *value) const
+{
+    rtlFailUnexpected();
+    return 0;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Unreachable?

File: rtl/eclrtl/rtlnewkey.cpp [hunk 20]
```diff
@@ -0,0 +1,2133 @@
...
+        if (c=='\\')
+        {
+            c = *in++;
+            switch (c)
+            {
+            case '\'':
+            case '\\':
+                break;
+            default:
+                UNIMPLEMENTED;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlnewkey.cpp]: Is there a Jira open to complete this work? Would it be easier just to do it now? Is there any existing code that we can common up to do it?

File: rtl/eclrtl/rtlnewkey.cpp [hunk 21]
```diff
@@ -0,0 +1,2133 @@
...
+
+
+void deserializeSet(ISetCreator * creator, const char * filter)
+{
+    while (*filter)
+    {
+        char startRange = *filter++;
+        if (startRange != '(' && startRange != '[')
+            throw MakeStringException(0, "Invalid filter string: expected [ or ( at start of range");
+        // Now we expect a constant - type depends on type of field. Assume string or int for now
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlnewkey.cpp]: Not sure this comment is really true...

File: rtl/eclrtl/rtlnewkey.cpp [hunk 22]
```diff
@@ -0,0 +1,2133 @@
...
+            //if (oneless(hival, t.getValue()))
+        }
+        return false;
+    }
+
+    //If this transition isn't shared then modify it directly, otherwise clone.  Always return a linked object.
+    ValueTransition * modifyTransition(const RtlTypeInfo & type, TransitionMask newMask)
+    {
+        assertex(value);
+        if (IsShared())
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlnewkey.cpp]: Any threading concerns doing this?

File: rtl/eclrtl/rtlnewkey.cpp [hunk 23]
```diff
@@ -0,0 +1,2133 @@
...
+
+bool SetFieldFilter::withinUpperRange(const RtlRow & row, unsigned range) const
+{
+    UNIMPLEMENTED;
+    return false;
+}
+
+bool SetFieldFilter::matches(const RtlRow & row, unsigned range) const
+{
+    UNIMPLEMENTED;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlnewkey.cpp]: Is there a Jira for completing these?

**Comment by richardkchapman** [comment]
Some general comments/questions:

1. Unreachable code after UNIMPLEMENTED/throwUnexpected() etc. This is a style question but I think our normal convention is to avoid it. Some compilers may warn (and I suspect others may warn the other way).
2. Javadoc comments on new classes/functions/interfaces - I think we agreed these should always be provided
3. Have you done any coverage analysis on your unit test runs? I found it very useful to do so when I was writing unit tests for the row translator.

Other than that, mostly minor comments inline above.

**Comment by richardkchapman** [comment]
@ghalliday

File: rtl/eclrtl/eclhelper_dyn.cpp (hunk 12 above)
Comment (ghalliday) [reply rtl/eclrtl/eclhelper_dyn.cpp]: I thought I would leave that for a separate PR

File: rtl/eclrtl/rtlfield.cpp (hunk 16 above)
Comment (ghalliday) [reply rtl/eclrtl/rtlfield.cpp]: It isn't - I will fix

File: rtl/eclrtl/rtlnewkey.cpp (hunk 22 above)
Comment (ghalliday) [reply rtl/eclrtl/rtlnewkey.cpp]: No - access to these sets needs to be single threaded

File: rtl/eclrtl/rtlfield.cpp [hunk 24]
```diff
@@ -339,12 +384,23 @@ size32_t RtlIntTypeInfo::build(ARowBuilder &builder, size32_t offset, const RtlF
...
-    builder.ensureCapacity(length+offset, field->name);
+    builder.ensureCapacity(length+offset, queryName(field));
     rtlWriteInt(builder.getSelf() + offset, val, length);
     offset += length;
     return offset;
 }
 
+size32_t RtlIntTypeInfo::buildString(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, size32_t size, const char *value) const
+{
+    return buildInt(builder, offset, field, rtlStrToInt8(size, value));
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Should this take any account of the signedness of the target field?

File: rtl/eclrtl/rtlfield.cpp [hunk 25]
```diff
@@ -791,13 +917,14 @@ void RtlStringTypeInfo::getString(size32_t & resultLen, char * & result, const v
 
 void RtlStringTypeInfo::getUtf8(size32_t & resultLen, char * & result, const void * ptr) const
 {
+    //MORE: What about isEbcdic?
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Should be implemented (or JIRA opened) really...

File: rtl/eclrtl/rtlfield.cpp [hunk 26]
```diff
@@ -1422,13 +1625,33 @@ double RtlDecimalTypeInfo::getReal(const void * ptr) const
...
 
 size32_t RtlCharTypeInfo::build(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, IFieldSource &source) const
 {
     throwUnexpected();  // Can't have a field of type char
 }
 
+size32_t RtlCharTypeInfo::buildString(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, size32_t size, const char *value) const
+{
+    rtlFailUnexpected();
+    return 0;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Unreachable

File: rtl/eclrtl/rtlfield.cpp [hunk 27]
```diff
@@ -2090,8 +2443,25 @@ __int64 RtlRecordTypeInfo::getInt(const void * ptr) const
...
+{
+    return compareFields(fields, left, right);
+}
+
 //-------------------------------------------------------------------------------------------------------------------
 
+size32_t RtlCompoundTypeInfo::buildString(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, size32_t size, const char *value) const
+{
+    UNIMPLEMENTED;
+    return 0;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Unreachable

File: rtl/eclrtl/rtlfield.cpp [hunk 28]
```diff
@@ -2090,8 +2443,25 @@ __int64 RtlRecordTypeInfo::getInt(const void * ptr) const
...
+}
+
+size32_t RtlCompoundTypeInfo::buildUtf8(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, size32_t len, const char *value) const
+{
+    UNIMPLEMENTED;
+    return 0;
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlfield.cpp]: Unreachable

File: rtl/eclrtl/rtlnewkey.cpp (hunk 22 above)
Comment (richardkchapman) [reply rtl/eclrtl/rtlnewkey.cpp]: At create time, I can accept (and this is create time I think). I hope they can be accessed from multiple threads after created though!

File: rtl/eclrtl/rtlfield.cpp (hunk 24 above)
Comment (ghalliday) [reply rtl/eclrtl/rtlfield.cpp]: I don't think it needs to string to int should work for (-2^(n-1)..2^n-1

**Pull request merged**